ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/task/RetryingGetBlocksFromPeersTask.java

```diff
@@ -0,0 +1,112 @@
+
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.CompletableFuture;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class RetryingGetBlocksFromPeersTask
+    extends AbstractRetryingSwitchingPeerTask<PeerTaskResult<List<Block>>> {
+
+  private static final Logger LOG = LoggerFactory.getLogger(RetryingGetBlocksFromPeersTask.class);
+
+  private final ProtocolSchedule protocolSchedule;
+  private final List<BlockHeader> headers;
+
+  protected RetryingGetBlocksFromPeersTask(
+      final EthContext ethContext,
+      final ProtocolSchedule protocolSchedule,
+      final MetricsSystem metricsSystem,
+      final int maxRetries,
+      final List<BlockHeader> headers) {
+    super(ethContext, metricsSystem, Objects::isNull, maxRetries);
+    this.protocolSchedule = protocolSchedule;
+    this.headers = headers;
+  }
+
+  public static RetryingGetBlocksFromPeersTask forHeaders(
+      final ProtocolSchedule protocolSchedule,
+      final EthContext ethContext,
+      final MetricsSystem metricsSystem,
+      final int maxRetries,
+      final List<BlockHeader> headers) {
+    return new RetryingGetBlocksFromPeersTask(
+        ethContext, protocolSchedule, metricsSystem, maxRetries, headers);
+  }
+
+  @Override
+  protected CompletableFuture<PeerTaskResult<List<Block>>> executeTaskOnCurrentPeer(
+      final EthPeer currentPeer) {
+    final GetBodiesFromPeerTask getBodiesTask =
+        GetBodiesFromPeerTask.forHeaders(
+            protocolSchedule, getEthContext(), headers, getMetricsSystem());
+    getBodiesTask.assignPeer(currentPeer);
+
+    return executeSubTask(getBodiesTask::run)
+        .thenApply(
+            peerResult -> {
+              debugLambda(
+                  LOG,
+                  "Got {} blocks {} from peer {}, attempt {}",
+                  peerResult.getResult()::size,
+                  peerResult.getPeer()::toString,
+                  this::getRetryCount);
+
+              if (peerResult.getResult().isEmpty()) {
+                currentPeer.recordUselessResponse("GetBodiesFromPeerTask");
+                throw new IncompleteResultsException(
+                    "No blocks returned by peer " + currentPeer.getShortNodeId());
+              }
+
+              result.complete(peerResult);
+              return peerResult;
+            });
+  }
+
+  @Override
+  protected boolean isRetryableError(final Throwable error) {
+    return super.isRetryableError(error) || error instanceof IncompleteResultsException;
+  }
+
+  @Override
+  protected void handleTaskError(final Throwable error) {
+    if (getRetryCount() < getMaxRetries()) {
+      debugLambda(
+          LOG,
+          "Failed to get {} blocks from peer {}, attempt {}, retrying later",
+          headers::size,
+          this::getAssignedPeer,
+          this::getRetryCount);
+    } else {
+      LOG.debug("Failed to get {} blocks after {} retries", headers.size(), getRetryCount());
```

Comment:
should the peer be logged here too

Reply:
This is the case when all peers have failed, for this the peer is not logged